### PR TITLE
NIFI-12885 Add convenience method to `Record` to parse DateTime objects

### PR DESCRIPTION
**Pull request by knguyen1**
# Summary

[NIFI-12885](https://issues.apache.org/jira/browse/NIFI-12885)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [X] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Review by exceptionfactory — COMMENTED** [review]
Thanks for putting this together @knguyen1.

The general approach looks good on initial review. I will take a closer look soon.

Can you revert the import reordering changes? Thanks!

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
@knguyen1 The test failures on MacOS and Windows indicate a problem with the expectations for timestamp conversion across time zones.

It can be difficult to set proper expectations that will work regardless of local machine time zone, but it is very important to get these tests correct to avoid confusing runtime problems.

For example, converting from OffsetDateTime to Instant and then comparing the values is not an ideal approach due to time zone differences.

It may be helpful to scope down the number of timestamps in the `provideTimestamps` method to focus on differences. It is not necessary to exercise every variation of formatting, but it is more important to exercise different types of time fields in terms of year, month, day, hour, minute, second and time zone.

**Comment by knguyen1** [comment]
> It may be helpful to scope down the number of timestamps in the provideTimestamps method to focus on differences.

@exceptionfactory  Okay.  I scoped it down to 2 formats: `ISO_LOCAL_DATE_TIME` and `ISO_OFFSET_DATE_TIME` and I limited the offset to `+00:00`.  Because there were parsing failures previously, I think it's indicative of some race conditions we're not covering, which is out of scope for this change.  Given that best practice is ISO format and UTC+0 for most db's anyway, I think we should be fine.

**Comment by exceptionfactory** [comment]
> > It may be helpful to scope down the number of timestamps in the provideTimestamps method to focus on differences.
> 
> @exceptionfactory Okay. I scoped it down to 2 formats: `ISO_LOCAL_DATE_TIME` and `ISO_OFFSET_DATE_TIME` and I limited the offset to `+00:00`. Because there were parsing failures previously, I think it's indicative of some race conditions we're not covering, which is out of scope for this change. Given that best practice is ISO format and UTC+0 for most db's anyway, I think we should be fine.

Thanks for making the updates, unfortunately the tests are still failing:

```
Error:  org.apache.nifi.serialization.record.TestMapRecord.testGettingLocalDateAndOffsetDateTime(Object, String, long, long)[1] -- Time elapsed: 0.031 s <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <1641040496789> but was: <1641072896789>
```

The problem is not a race condition, but the fact of the different GitHub Runners have different local time zones. Converting between UTC and local time changes the expected epoch values, so the expectations need to be written such that they work across all time zones.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/MapRecord.java [hunk 1]
```diff
@@ -279,10 +282,27 @@ public Boolean getAsBoolean(final String fieldName) {
...
+        try {
+            // First, try to parse the data as if it can be converted to `LocalDateTime` then set `ZoneOffset.UTC`
+            final FieldConverter<Object, LocalDateTime> localDateTimeConverter = StandardFieldConverterRegistry.getRegistry().getFieldConverter(LocalDateTime.class);
+            final LocalDateTime localDateTime = localDateTimeConverter.convertField(getValue(fieldName), Optional.ofNullable(format), fieldName);
+            return OffsetDateTime.of(localDateTime, ZoneOffset.UTC);
+        } catch (DateTimeParseException exc) {
+            // Parsing fails if offset is not provided, error: `Unable to obtain OffsetDateTime from TemporalAccessor`
+            final FieldConverter<Object, OffsetDateTime> offsetDateTimeConverter = StandardFieldConverterRegistry.getRegistry().getFieldConverter(OffsetDateTime.class);
+            return offsetDateTimeConverter.convertField(getValue(fieldName), Optional.ofNullable(format), fieldName);
+        }
```
Comment (exceptionfactory) [inline nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/MapRecord.java:305]: The try-catch approach to parsing is not optimal from a performance perspective. There is a more general date parsing method that returns a `TemporalAccessor`, which can be evaluated for various time fields. It looks like this method might become more complicated in order to support a wider variety of string formats.

On closer consideration, however, throwing an exception may be more appropriate. Requesting an `OffsetDateTime` implies that the timezone is available in source field. Returning an `OffsetDateTime` with `UTC` can create confusion in calling functions, so this does not seem like the best approach.

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
Thanks for the recent updates @knguyen1. Taking a closer look at `getOffsetDateTime()` highlights a potential problem with the helper function doing too much. It seems better to throw an exception, as opposed to defaulting the time zone. This raises a wider question about the usefulness of this particular helper method, as opposed to putting the responsibility back on the caller to know the appropriate time class to use.

File: nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/MapRecord.java (hunk 1 above)
Comment (knguyen1) [reply nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/record/MapRecord.java:305]: Okay, I agree with this.  Will remove the UTC default.  Caller should call the appropriate helper method; throwing an exception is appropriate.

**Comment by knguyen1** [comment]
@exceptionfactory Hopefully we are close.  I added

1. `testGettingLocalDate`
2. `testGettingLocalDateTime`
3. `testGettingOffsetDateTime`

in their own tests.  In addition, the `Record` interface has an additional method `getAsLocalDateTime`.

Tests should now pass regardless of system default timezones.  I checked against `America/New_York`, `UTC+00`, etc.

**Review by exceptionfactory — APPROVED** [review]
Thanks for working through the feedback @knguyen1, the latest version looks good, and removes the ambiguity around the legacy `getAsDate()` method. +1 merging

**Pull request merged**